src/os/ObjectStore.h

```diff
@@ -1718,22 +1718,127 @@ class ObjectStore {
    * @param op_flags is CEPH_OSD_OP_FLAG_*
    * @returns number of bytes read on success, or negative error code on failure.
    */
-   virtual int read(
-    const coll_t& cid,
+  virtual int read(
+   const coll_t& cid,
+   const ghobject_t& oid,
+   uint64_t offset,
+   size_t len,
+   bufferlist& bl,
+   uint32_t op_flags = 0) = 0;
+  virtual int read(
+    CollectionHandle &c,
     const ghobject_t& oid,
     uint64_t offset,
     size_t len,
     bufferlist& bl,
-    uint32_t op_flags = 0) = 0;
-   virtual int read(
-     CollectionHandle &c,
-     const ghobject_t& oid,
-     uint64_t offset,
-     size_t len,
-     bufferlist& bl,
-     uint32_t op_flags = 0) {
-     return read(c->get_cid(), oid, offset, len, bl, op_flags);
-   }
+    uint32_t op_flags = 0) {
+    return read(c->get_cid(), oid, offset, len, bl, op_flags);
+  }
+
+  /*********************************
+   * ReadTransaction
+   *
+   * A Transaction represents a sequence of primitive non-mutating
+   * operations over the same ghobject in the same collection.
+   */
+  class ReadTransaction {
+  public:
+    virtual ~ReadTransaction() = default;
+
+    /**
+     * read -- read a byte range of data from an object.
+     *
+     * The operation can be performed synchronously or asynchronously
+     * solely depnding on the implementation's decision.
+     *
+     * Note: if reading from an offset past the end of the object, we
+     * return 0 (not, say, -EINVAL).
+     *
+     * @param offset location offset of first byte to be read
+     * @param len number of bytes to be read
+     * @param op_flags is CEPH_OSD_OP_FLAG_*
+     * @param destbl output bufferlist
+     * @param on_complete callback to be called
+     * @returns number of bytes read on success,
+     *          or negative error code on failure,
+     *          or -EINPROGESS if implementations decided to go async.
+     */
+    virtual int read(
+      uint64_t offset,
+      uint64_t length,
+      uint32_t flags,
+      ceph::bufferlist& destbl,
+      Context* on_complete) = 0;
```

Comment:
I'm not a big fan of using callbacks here. The approach usually imposes dynamic allocation of `Context` which is completely unnecessary if implementation decides to go synchronously (i.e. the cache hit scenario). However, the async machinery of `PrimaryLogPG` is already organized that way (`C_ChecksumRead`, `C_ExtentCmpRead`, `FillInVerifyExtent` and so on).

I believe ultimately we'll want to unify the paths (e.g. no special code for the `is_erasure()` case) and replace callbacks with futures, promises & lambdas. However, taking `std::function` wouldn't be free of the allocations as a closure may vary in size. Templating a virtual method (around functor type) is an obvious no-no. What we could do for the SeaStar-makeshift time is to introduce a dumb future-alike return type that converts lambda into `Context*` only when truly necessary.

```cpp
template <class ProdT, class ErrorT = int>
struct ret_t {
  // would boost::variant be an overkill here?
  ErrorT error;
  // xor
  boost::optional<ProdT> product;
  Context** async_context_dest;

  template <class F>
  void then(F&& f) {
    // handle error: throw it?
    // ...

    if (product) {
      // the sync scenario - no dynalloc, execute the functor immediately
      f(std::move(*product));
    } else {
      // not available yet -- async
      *async_context_dest = new LambdaContext(std::forward<F>(f));
    }
  }
};
```

Allowing for direct use of something like that would be a headache as `async_context_dest` might become a dangling pointer very easily (e.g. second call to `::read` invalidates instance from the first one). Enforcing implementations to guarantee it won't happen is even worse. However, encapsulating everything inside of `ObjectStore::ReadTransaction` might help: `::read()` returns `ret_t` but becomes non-`public`. What the client sees is `read_maybe_async` which would be
* non-virtual,
* templated around the functor type,
* defined at the level of `ObjectStore`,
* shared between `FileStore`, `BlueStore` & co.

The open question is: does the benefits justify the extra complexity?